### ISSUE TITLE
Get Magisk Working on Android 11-based GSI's

### ISSUE THREAD

**Issue by dragonBall561** [topjohnwu/Magisk#3593]
Hello! I have a Moto G8+ running a Lineage OS 18.1 GSI. Magisk doesn't work. Running v21 goes to bootloader on boot and v20 gets stuck in a bootloop.

I have successfully removed Magisk after each attempt. Could anyone please figure out what is going wrong?

**Comment by Displax** [comment]
https://github.com/topjohnwu/Magisk/issues/844

**Comment by dragonBall561** [comment]
Is there any fix?

**Comment by dragonBall561** [comment]
Also what about this? 
https://mobile.twitter.com/topjohnwu/status/1008056655001468928?lang=en

I have a phh build. (Andy Yan)

**Comment by dragonBall561** [comment]
Edit: Got working, flash successful. But magisk manager says "Installed: N/A" and won't last me access modules.

**Comment by yukosky** [comment]
:/

**Comment by dragonBall561** [comment]
I figured it out :)

How I did it:

**In TWRP**
1. Mount System
1. Make The Directory `/system_root/sbin`
1. Install Magisk **v21.1!!!**
1. Reboot System

**In System**
1. Get a Terminal Emulator (Or `adb root` then `adb shell`, but make sure you allow root debugging in settings)
1.  If using a Terminal Emulator, type `su`. Magisk will ask for root access, allow.
1. Type `echo $PATH`, and copy output.
1. Use a text editor that supports editing root files, or a terminal-based text editor if you are comfortable with one (nano, vim, etc.) to edit `/init.environ.rc`
1. Edit it to look like this:
```
# Stuff Here!!!
on early-init
      export PATH /sbin:(echo $PATH output here)
      # Stuff Here!!!
```
1. Reboot

**Comment by hwti** [comment]
@dragonBall561, thank you for the workaround

In my case, before creating `/sbin`, I had :
```
[   13.331584] init: cannot execv('/dev/Thu8xtl/magisk'). See the 'Debugging init' section of init's README.md for tips: Permission denied
```
So something doesn't work with the new installation path.
Once the system is booted, the `/dev/Thu8xtl/magisk` can be executed, so I don't know what is the issue.

Then the `PATH` change allows `/sbin/su` (Magisk) to be used instead of `/system/xbin/su` (phh).

**Comment by dragonBall561** [comment]
If you have ANY su installed, uninstall it. Make sure to create /sbin **before** flashing Magisk.

Also make sure you flash **v21.1** because older versions bootloop

**Comment by hwti** [comment]
It seems there are two issues :
 - There is a bind-mount (by Magisk ?) of the original `/system/bin/resetprop` to `/system/bin/magisk` (target of `resetprop` symlink) 
 - phh-su copies itself to `/dev/su/su`, and does a bind-mount to `/system/bin/su`, which "replaces" the `magisk` binary too

So running `phh-securize.sh` (which requires a GSI able to be remounted rw, and needs to be adapted to Magisk >=21 depending on when we launch the script) to remove the builtin su fixes the second problem.

But the resetprop issue must be fixed first, else I lose root. Removing `/system/bin/resetprop` works, but it breaks the ROM use of the command in boot scripts (I don't know why the current Magisk resetprop doesn't seem to work).
Renaming `/system/bin/resetprop` to `phh-resetprop`, and adjusting `/system/bin/rw-system.sh` and `/system/bin/phh-prop-handler.sh` works.

**Comment by dragonBall561** [comment]
Is there any way to **completely** remove phh su. You can't have phh su with magisk, it does weird stuff.

**Comment by dragonBall561** [comment]
Thanks I'll add it to the guide :)

**Comment by hwti** [comment]
@dragonBall561 You are doing too many things.

With the `phh-securize.sh` and the renaming to `phh-resetprop`, Magisk can be installed in the default Android 11 location, there is no need to create `/sbin` and modify `PATH` any more.

Running `phh-securize.sh` before installing Magisk is installed like you suggested is actually easier that what I had to do manually, but it requires TWRP (and for example people upgrading from an Android 10 GSI and keeping their patched boot image).

Ideally, Magisk shouldn't overwrite itself when there is a resetprop in the ROM and phh-su should avoid overwriting Magisk. Then the phh-securize.sh could be improved to work even when Magisk runs over phh-su, like I did for the previous install location.

**Comment by dragonBall561** [comment]
Ahh ok. I thought there was an extra issue. I didn't realize that you were just suggesting another way :) I will include that.

**Comment by lss4** [comment]
> With the `phh-securize.sh` and the renaming to `phh-resetprop`, Magisk can be installed in the default Android 11 location, there is no need to create `/sbin` and modify `PATH` any more.

From what I tested on my Razer Phone 2 with Nusantara GSI (which is already securized), renaming to `phh-resetprop` is indeed mandatory but I'm still not sure about the creating of `/sbin` (I did anyway, and Magisk works).

> Running `phh-securize.sh` before installing Magisk is installed like you suggested is actually easier that what I had to do manually, but it requires TWRP (and for example people upgrading from an Android 10 GSI and keeping their patched boot image).

Haven't tested this part as from my own experience, major upgrades from GSI to GSI is not straightforward and I usually just back things up with Migrate, make a clean flash, then restore the stuffs I have backed up once I verified everything important is working correctly.

> Ideally, Magisk shouldn't overwrite itself when there is a resetprop in the ROM and phh-su should avoid overwriting Magisk. Then the phh-securize.sh could be improved to work even when Magisk runs over phh-su, like I did for the previous install location.

In case I really cannot do anything to the system partition (such as with the new [Dynamic Partitioning](https://source.android.com/devices/tech/ota/dynamic_partitions/implement)), it's still necessary to look for ways to at least coexist with phh's `resetprop`.

**Comment by nickelnine** [comment]
Appreciate the inputs from everyone but it's hard to follow for a noob like me. Can someone kindly summarize the steps required to flash Magisk on an Android 11 GSI?

**Comment by Skorpion96** [comment]
i am on an android 13 gsi with no sbin and no twrp, also my boot.img doesn't have a ramdisk, flashing magisk on recovery doesn't work neither, if i do the procedure and release buttons at logo i get into recovery anyway, could you please provide instructions on how to install magisk into the system partition? I tried as well magisk delta system install but it doesn't work on f2fs

**Comment by Skorpion96** [comment]
may i say this is working on gsi with magisk 25.2:https://github.com/HuskyDG/magisk-files/issues/118#issuecomment-2381384721 i just had to add on init.rc a service to set selinux permissive, i'll grab everything 
[Init_magisk.txt](https://github.com/user-attachments/files/17202866/Init_magisk.txt)
